core/platform/auth/firebase_auth_services.py

```diff
@@ -70,79 +71,63 @@
 
 transaction_services = models.Registry.import_transaction_services()
 
+# Session cookies only provide temporary authentication, so they are expected to
+# become obsolete over time. The following errors is the full list of situations
+# this happens.
+EXPECTED_SESSION_COOKIE_EXCEPTIONS = (
+    firebase_auth.InvalidSessionCookieError,
+    firebase_auth.ExpiredSessionCookieError,
+    firebase_auth.RevokedSessionCookieError,
+)
 
-@contextlib.contextmanager
-def _acquire_firebase_context():
-    """Returns a context for calling the Firebase Admin SDK."""
-    app = firebase_admin.initialize_app()
-    try:
-        yield
-    finally:
-        if app is not None:
-            firebase_admin.delete_app(app)
 
+def establish_auth_session(request, response):
+    """Sets login cookies to maintain a user's sign-in session.
 
-def _verify_id_token(auth_header):
-    """Verifies whether the auth_header has a valid Firebase-provided ID token.
+    Args:
+        request: webapp2.Request. The request with the authorization to begin a
+            new session.
+        response: webapp2.Response. The response to establish the new session
+            upon.
+    """
+    cookie_claims = (
+        _get_auth_claims_from_session_cookie(_get_session_cookie(request)))
 
-    Oppia's authorization headers use OAuth 2.0's Bearer authentication scheme.
+    # If the request already contains a valid session cookie, then there's no
+    # action necessary; the session is already established.
+    if cookie_claims is not None:
+        return
 
-    Bearer authentication (a.k.a. token authentication) is an HTTP
-    authentication scheme based on "bearer tokens", an encrypted JWT generated
-    by a trusted identity provider in response to login requests.
+    with _firebase_admin_context():
+        fresh_cookie = firebase_auth.create_session_cookie(
+            _get_id_token(request), feconf.FIREBASE_SESSION_COOKIE_MAX_AGE)
 
-    The name "Bearer authentication" can be understood as: "give access to the
-    bearer of this token." These tokens _must_ be sent in the `Authorization`
-    header of HTTP requests, and _must_ have the format: `Bearer <token>`.
+    response.set_cookie(
+        feconf.FIREBASE_SESSION_COOKIE_NAME, value=fresh_cookie,
+        max_age=feconf.FIREBASE_SESSION_COOKIE_MAX_AGE,
+        httponly=True, overwrite=True, secure=not constants.DEV_MODE)
```

Comment:
Why httponly?

Also, mild preference for putting parens around "not constants.DEV_MODE", but this is optional. I also thought @vojtechjelinek was refactoring this to have DEV_MODE, PROD_MODE and EMULATOR_MODE -- so just wanted to check that you've considered those cases? /cc @vojtechjelinek for a quick review of this part.